File: src/py_back/manager.py
import langs
from flask import Flask
from flask import request, render_template_string
import os

web = Flask(__name__)         

def text_content(path: str) -> str:
     '''
        Give content of a file entire
        only text without escape chars
     '''
     content = ''
     
     for line in open(path, 'r').readlines():
        content += line.replace('\n', '')

     return content   
# Define thory web logic in Flask
class Tables():
    def __init__(self, tab: str):
        '''
            Make easier GTK window with tables to
            show data, a very common kind of
            programs for enterprises:

            Tables.table, text for the table

            Tables.option, text with selected option from parametric input
            
            Tables.row, list with the last input has a length of Tables.entries as max

            Tables.external, external function for cleaner and better envs
            
            Tables.entries, number of input for the interface
        '''
        # Object properties
        self.table: str = tab

        self.row: list[str] = []

        self.option: str = ''
        
        self.entries: int = 4

        self.external: None = int(0)

    @web.route('/sended')   
    def get_inputs(self):
        '''
            Get forms values from cookies
            using flask
        '''
        cookie = request.cookies.get('thory').split(',')
        # Update currency name
        self.option = cookie[2]
        # Delete data that don't need now
        cookie.pop(2)
        # Get values from imputs excluding currency name
        for input_data in cookie:
            # Get sended by the page and add to row data
            self.row.append(input_data)
        return render_template_string(text_content('index.html'), result = '')

    @web.route('/')                
    def last_outputs(self, outputs: list[str]):
        '''
            Give text to display
            in the end send to 
            the web page using flask
        '''
        self.run()
        
        content = list(text_content('index.html'))

        counter = 1
         
        for output in outputs:
            # Insert from last label to first label
            content.insert((content.index('</label>') - counter), f'<h1>{output}</h1>')
            counter += 1
        # Insert inside <table> tag
        content.insert((content.index('</table>') - 1), self.table)
            
        return content.__str__()
                     
    def update(self):
        '''
            Send data and update web, show result
        '''
        # Get the last inputs from the form
        self.get_inputs()
        # Make the table output
        text = ''
        # Get all the input and format from the row list
        for data in self.row:
            # Add multiple dash as seperator on each row as chars as the first row
            text += f'|\t{data}\t'
        # Update table with new data
        self.table += f'\n{"-" * (self.row_length + text.__len__())}\n{text}|\n'
        del text
        
    def send_data(self):
            '''
                Event when the data are loaded
                all and the button is pressed
                for don't show weird text by
                missed data not written yet.
            '''        
            # Update view for show the result and save
            self.external()
            # Restart for load more products
            self.row: list[str] = []

            self.get_inputs()
        
    def run(self):
         '''
            Show the current graphical
            interface
         '''
         # Send data to page with response JSON
         os.system(f'flask {__name__} --app run')

File: src/manager.rs
pub struct Tables
{
    // Atributes
    priv table : &str;
    priv option : &str;
    priv row : [&str; 3];
    priv size : [&str; 3];
    priv ins_container : (&str);
    priv outs_container : (&str);
    priv entries : u16;
    priv changes : u8;
}
// Methods
impl Tables
{
    pub Tables(&self, tab : &str, window_size : [&str; 3])
    {
       // Init conf for dash  
        self.table = tab;
        self.option = "";
        self.row = ["", "", ""];
        self.changes = 0;
        self.entries = 3;
      // Create window of that size
        self.size = window_size;
    }

    pub get_table(&self) -> &str
    {
        return self.table;
    }

    pub get_row(&self) -> [&str; 3]
    {
        return self.row;
    }

    pub get_option(&self) -> &str
    {
        return self.option;
    }

    pub get_changes(&self) -> u8
    {
        return self.changes;
    }

    pub get_entries(&self) -> u16
    {
        return self.entries;
    }

    pub set_inputs(&self, list : &str)
    {
        for input in list
        {
            // Add input to ins_container
            self.ins_container.push(input);
        }

        self.entries = list.len();
    }

    pub run(&self)
    {
        // Create interface
        slint :: slint!
        {
            export component Dash
            {
                inherits Window
                {
                   // Add multiple inputs 
                    for entry in self.ins_container
                    {
                        Text
                        {
                            text: entry;
                            color: blue;
                        }
                        Entry
                        {
                        }
                    }
                    // Add sender button
                    Button
                    {
                        label: "Send";
                    }
                    // Show table with results
                    Text
                    {
                        text: self.table;
                        color: black;
                    }
                }
            }
        }    
    }

    pub parametric_input(&self, paramaters : (&str), property_name : &str)
    {
        // Selection option menu input
    }

    pub on_option_changed(&self)
    {
        // Load new option from the menu
    }

    pub update_table(&self)
    {
        // Update table
        self.table = format!("{}\n", self.table);
        for input in self.row
        {
            self.table = format!("{}\t{}", self.table, input)
        }
    }

    pub send_data(&self)
    {
        // When get all the entries changed then update table and refresh interface
        if (self.changes == self.entries)
        {
            self.update_table()
        }
        else
        {
            // Restart changes
            self.changes = 0;
        }
    }
}